### ISSUE TITLE
Error: retrieving `contact` for KeyVault: keyvault.BaseClient#GetCertificateContacts: Failure sending request: StatusCode=0 -- Original Error: context deadline exceeded

### ISSUE THREAD

**Issue by marrobi** [terraform-providers/terraform-provider-azurerm#10501]
### Terraform Version

```
➜  ~ terraform --version
Terraform v0.14.6
```

### Terraform Configuration Files

```terraform
terraform {
  required_providers {
    azurerm = {
      source  = "hashicorp/azurerm"
      version = "=2.46.0"
    }
  }
}

provider "azurerm" {
  features {}
}

data "azurerm_subscription" "current" {
}

resource "azurerm_resource_group" "core" {
  location = "uksouth"
  name     = "tmp-core"
}

resource "azurerm_virtual_network" "core" {
  name                = "vnet-core"
  resource_group_name = azurerm_resource_group.core.name
  location            = azurerm_resource_group.core.location
  address_space       = ["10.0.0.0/24"]
}

resource "azurerm_subnet" "sharedServices" {
  name                                           = "sharedServicesSubnet"
  virtual_network_name                           = azurerm_virtual_network.core.name
  resource_group_name                            = azurerm_resource_group.core.name
  address_prefixes                               = ["10.0.0.0/24"]
  enforce_private_link_endpoint_network_policies = true
}

resource "azurerm_key_vault" "core" {
  name                            = "kv-core-uniquename"
  location                        = azurerm_resource_group.core.location
  enabled_for_deployment          = false
  enabled_for_template_deployment = true
  enabled_for_disk_encryption     = false
  soft_delete_retention_days      = 7
  tenant_id                       = data.azurerm_subscription.current.tenant_id
  sku_name                        = "standard"
  resource_group_name             = azurerm_resource_group.core.name

}

resource "azurerm_private_endpoint" "keyvault-private-endpoint" {
  name                = "pe-kv-core"
  resource_group_name = azurerm_resource_group.core.name
  location            = azurerm_resource_group.core.location
  subnet_id           = azurerm_subnet.sharedServices.id
  private_service_connection {
    private_connection_resource_id = azurerm_key_vault.core.id
    name                           = "pe-kv-core"
    subresource_names              = ["vault"]
    is_manual_connection           = false
  }
  private_dns_zone_group {
    name                 = "privatelink.vaultcore.azure.net"
    private_dns_zone_ids = [azurerm_private_dns_zone.keyvault.id]
  }
}

resource "azurerm_private_dns_zone" "keyvault" {
  name                = "privatelink.vaultcore.azure.net"
  resource_group_name = azurerm_resource_group.core.name

}
```

### Debug Output
```
2021-02-07T16:25:18.140Z [DEBUG] plugin.terraform-provider-azurerm_v2.46.0_x5: [DEBUG] AzureRM Request: 
2021-02-07T16:25:18.140Z [DEBUG] plugin.terraform-provider-azurerm_v2.46.0_x5: GET /certificates/contacts?api-version=2016-10-01 HTTP/1.1
2021-02-07T16:25:18.140Z [DEBUG] plugin.terraform-provider-azurerm_v2.46.0_x5: Host: kv-core-xxxxxx.vault.azure.net
2021-02-07T16:25:18.140Z [DEBUG] plugin.terraform-provider-azurerm_v2.46.0_x5: User-Agent: Go/go1.15.5 (amd64-linux) go-autorest/v14.2.1 Azure-SDK-For-Go/v49.2.0 keyvault/2016-10-01 HashiCorp Terraform/0.14.6 (+https://www.terraform.io) Terraform Plugin SDK/1.13.1 terraform-provider-azurerm/2.46.0 pid-222c6c49-1b0a-5959-a213-6608f9eb8820
2021-02-07T16:25:18.140Z [DEBUG] plugin.terraform-provider-azurerm_v2.46.0_x5: Accept-Encoding: gzip
2021-02-07T16:25:18.140Z [DEBUG] plugin.terraform-provider-azurerm_v2.46.0_x5: 
2021-02-07T16:25:18.140Z [DEBUG] plugin.terraform-provider-azurerm_v2.46.0_x5: 
2021-02-07T16:25:18.316Z [DEBUG] plugin.terraform-provider-azurerm_v2.46.0_x5: [DEBUG] Request to https://kv-core-xxxxxx.vault.azure.net/certificates/contacts?api-version=2016-10-01 completed with no response
2021/02/07 16:25:19 [TRACE] dag/walk: vertex "meta.count-boundary (EachMode fixup)" is waiting for "azurerm_private_endpoint.keyvault-private-endpoint (expand)"
2021/02/07 16:25:19 [TRACE] dag/walk: vertex "provider[\"registry.terraform.io/hashicorp/azurerm\"] (close)" is waiting for "azurerm_private_endpoint.keyvault-private-endpoint (expand)"
2021/02/07 16:25:19 [TRACE] dag/walk: vertex "root" is waiting for "meta.count-boundary (EachMode fixup)"
2021/02/07 16:25:22 [TRACE] dag/walk: vertex "azurerm_private_endpoint.keyvault-private-endpoint (expand)" is waiting for "azurerm_key_vault.core (expand)"
2021/02/07 16:25:24 [TRACE] dag/walk: vertex "root" is waiting for "meta.count-boundary (EachMode fixup)"
2021/02/07 16:25:24 [TRACE] dag/walk: vertex "provider[\"registry.terraform.io/hashicorp/azurerm\"] (close)" is waiting for "azurerm_private_endpoint.keyvault-private-endpoint (expand)"
2021/02/07 16:25:24 [TRACE] dag/walk: vertex "meta.count-boundary (EachMode fixup)" is waiting for "azurerm_private_endpoint.keyvault-private-endpoint (expand)"
2021/02/07 16:25:27 [TRACE] dag/walk: vertex "azurerm_private_endpoint.keyvault-private-endpoint (expand)" is waiting for "azurerm_key_vault.core (expand)"
2021/02/07 16:25:29 [TRACE] dag/walk: vertex "meta.count-boundary (EachMode fixup)" is waiting for "azurerm_private_endpoint.keyvault-private-endpoint (expand)"
2021/02/07 16:25:29 [TRACE] dag/walk: vertex "root" is waiting for "meta.count-boundary (EachMode fixup)"
2021/02/07 16:25:29 [TRACE] dag/walk: vertex "provider[\"registry.terraform.io/hashicorp/azurerm\"] (close)" is waiting for "azurerm_private_endpoint.keyvault-private-endpoint (expand)"
2021/02/07 16:25:32 [TRACE] dag/walk: vertex "azurerm_private_endpoint.keyvault-private-endpoint (expand)" is waiting for "azurerm_key_vault.core (expand)"
2021/02/07 16:25:34 [TRACE] dag/walk: vertex "provider[\"registry.terraform.io/hashicorp/azurerm\"] (close)" is waiting for "azurerm_private_endpoint.keyvault-private-endpoint (expand)"
2021/02/07 16:25:34 [TRACE] dag/walk: vertex "root" is waiting for "meta.count-boundary (EachMode fixup)"
2021/02/07 16:25:34 [TRACE] dag/walk: vertex "meta.count-boundary (EachMode fixup)" is waiting for "azurerm_private_endpoint.keyvault-private-endpoint (expand)"
2021/02/07 16:25:37 [TRACE] dag/walk: vertex "azurerm_private_endpoint.keyvault-private-endpoint (expand)" is waiting for "azurerm_key_vault.core (expand)"
2021/02/07 16:25:39 [TRACE] dag/walk: vertex "root" is waiting for "meta.count-boundary (EachMode fixup)"
2021/02/07 16:25:39 [TRACE] dag/walk: vertex "provider[\"registry.terraform.io/hashicorp/azurerm\"] (close)" is waiting for "azurerm_private_endpoint.keyvault-private-endpoint (expand)"
2021/02/07 16:25:39 [TRACE] dag/walk: vertex "meta.count-boundary (EachMode fixup)" is waiting for "azurerm_private_endpoint.keyvault-private-endpoint (expand)"
2021/02/07 16:25:42 [TRACE] dag/walk: vertex "azurerm_private_endpoint.keyvault-private-endpoint (expand)" is waiting for "azurerm_key_vault.core (expand)"
2021/02/07 16:25:44 [TRACE] dag/walk: vertex "provider[\"registry.terraform.io/hashicorp/azurerm\"] (close)" is waiting for "azurerm_private_endpoint.keyvault-private-endpoint (expand)"
2021/02/07 16:25:44 [TRACE] dag/walk: vertex "root" is waiting for "meta.count-boundary (EachMode fixup)"
2021/02/07 16:25:44 [TRACE] dag/walk: vertex "meta.count-boundary (EachMode fixup)" is waiting for "azurerm_private_endpoint.keyvault-private-endpoint (expand)"
2021/02/07 16:25:47 [TRACE] dag/walk: vertex "azurerm_private_endpoint.keyvault-private-endpoint (expand)" is waiting for "azurerm_key_vault.core (expand)"
2021/02/07 16:25:49 [TRACE] dag/walk: vertex "meta.count-boundary (EachMode fixup)" is waiting for "azurerm_private_endpoint.keyvault-private-endpoint (expand)"
2021/02/07 16:25:49 [TRACE] dag/walk: vertex "provider[\"registry.terraform.io/hashicorp/azurerm\"] (close)" is waiting for "azurerm_private_endpoint.keyvault-private-endpoint (expand)"
2021/02/07 16:25:49 [TRACE] dag/walk: vertex "root" is waiting for "meta.count-boundary (EachMode fixup)"
2021/02/07 16:25:53 [TRACE] dag/walk: vertex "azurerm_private_endpoint.keyvault-private-endpoint (expand)" is waiting for "azurerm_key_vault.core (expand)"
2021/02/07 16:25:54 [TRACE] dag/walk: vertex "provider[\"registry.terraform.io/hashicorp/azurerm\"] (close)" is waiting for "azurerm_private_endpoint.keyvault-private-endpoint (expand)"
2021/02/07 16:25:54 [TRACE] dag/walk: vertex "root" is waiting for "meta.count-boundary (EachMode fixup)"
2021/02/07 16:25:54 [TRACE] dag/walk: vertex "meta.count-boundary (EachMode fixup)" is waiting for "azurerm_private_endpoint.keyvault-private-endpoint (expand)"
2021/02/07 16:25:58 [TRACE] dag/walk: vertex "azurerm_private_endpoint.keyvault-private-endpoint (expand)" is waiting for "azurerm_key_vault.core (expand)"
2021/02/07 16:25:59 [TRACE] dag/walk: vertex "meta.count-boundary (EachMode fixup)" is waiting for "azurerm_private_endpoint.keyvault-private-endpoint (expand)"
2021/02/07 16:25:59 [TRACE] dag/walk: vertex "root" is waiting for "meta.count-boundary (EachMode fixup)"
2021/02/07 16:25:59 [TRACE] dag/walk: vertex "provider[\"registry.terraform.io/hashicorp/azurerm\"] (close)" is waiting for "azurerm_private_endpoint.keyvault-private-endpoint (expand)"
2021/02/07 16:26:03 [TRACE] dag/walk: vertex "azurerm_private_endpoint.keyvault-private-endpoint (expand)" is waiting for "azurerm_key_vault.core (expand)"
2021/02/07 16:26:04 [TRACE] dag/walk: vertex "root" is waiting for "meta.count-boundary (EachMode fixup)"
2021/02/07 16:26:04 [TRACE] dag/walk: vertex "provider[\"registry.terraform.io/hashicorp/azurerm\"] (close)" is waiting for "azurerm_private_endpoint.keyvault-private-endpoint (expand)"
2021/02/07 16:26:04 [TRACE] dag/walk: vertex "meta.count-boundary (EachMode fixup)" is waiting for "azurerm_private_endpoint.keyvault-private-endpoint (expand)"
2021/02/07 16:26:08 [TRACE] dag/walk: vertex "azurerm_private_endpoint.keyvault-private-endpoint (expand)" is waiting for "azurerm_key_vault.core (expand)"
2021/02/07 16:26:09 [TRACE] dag/walk: vertex "meta.count-boundary (EachMode fixup)" is waiting for "azurerm_private_endpoint.keyvault-private-endpoint (expand)"
2021/02/07 16:26:09 [TRACE] dag/walk: vertex "root" is waiting for "meta.count-boundary (EachMode fixup)"
2021/02/07 16:26:09 [TRACE] dag/walk: vertex "provider[\"registry.terraform.io/hashicorp/azurerm\"] (close)" is waiting for "azurerm_private_endpoint.keyvault-private-endpoint (expand)"
2021/02/07 16:26:13 [TRACE] dag/walk: vertex "azurerm_private_endpoint.keyvault-private-endpoint (expand)" is waiting for "azurerm_key_vault.core (expand)"
2021/02/07 16:26:14 [TRACE] dag/walk: vertex "provider[\"registry.terraform.io/hashicorp/azurerm\"] (close)" is waiting for "azurerm_private_endpoint.keyvault-private-endpoint (expand)"
2021/02/07 16:26:14 [TRACE] dag/walk: vertex "root" is waiting for "meta.count-boundary (EachMode fixup)"
2021/02/07 16:26:14 [TRACE] dag/walk: vertex "meta.count-boundary (EachMode fixup)" is waiting for "azurerm_private_endpoint.keyvault-private-endpoint (expand)"
2021/02/07 16:26:18 [TRACE] dag/walk: vertex "azurerm_private_endpoint.keyvault-private-endpoint (expand)" is waiting for "azurerm_key_vault.core (expand)"
2021/02/07 16:26:19 [TRACE] dag/walk: vertex "meta.count-boundary (EachMode fixup)" is waiting for "azurerm_private_endpoint.keyvault-private-endpoint (expand)"
2021/02/07 16:26:19 [TRACE] dag/walk: vertex "provider[\"registry.terraform.io/hashicorp/azurerm\"] (close)" is waiting for "azurerm_private_endpoint.keyvault-private-endpoint (expand)"
2021/02/07 16:26:19 [TRACE] dag/walk: vertex "root" is waiting for "meta.count-boundary (EachMode fixup)"
2021/02/07 16:26:23 [TRACE] dag/walk: vertex "azurerm_private_endpoint.keyvault-private-endpoint (expand)" is waiting for "azurerm_key_vault.core (expand)"
2021/02/07 16:26:24 [TRACE] dag/walk: vertex "root" is waiting for "meta.count-boundary (EachMode fixup)"
2021/02/07 16:26:24 [TRACE] dag/walk: vertex "provider[\"registry.terraform.io/hashicorp/azurerm\"] (close)" is waiting for "azurerm_private_endpoint.keyvault-private-endpoint (expand)"
2021/02/07 16:26:24 [TRACE] dag/walk: vertex "meta.count-boundary (EachMode fixup)" is waiting for "azurerm_private_endpoint.keyvault-private-endpoint (expand)"
2021/02/07 16:26:28 [TRACE] dag/walk: vertex "azurerm_private_endpoint.keyvault-private-endpoint (expand)" is waiting for "azurerm_key_vault.core (expand)"
2021/02/07 16:26:29 [TRACE] dag/walk: vertex "root" is waiting for "meta.count-boundary (EachMode fixup)"
2021/02/07 16:26:29 [TRACE] dag/walk: vertex "meta.count-boundary (EachMode fixup)" is waiting for "azurerm_private_endpoint.keyvault-private-endpoint (expand)"
2021/02/07 16:26:29 [TRACE] dag/walk: vertex "provider[\"registry.terraform.io/hashicorp/azurerm\"] (close)" is waiting for "azurerm_private_endpoint.keyvault-private-endpoint (expand)"
2021/02/07 16:26:33 [TRACE] dag/walk: vertex "azurerm_private_endpoint.keyvault-private-endpoint (expand)" is waiting for "azurerm_key_vault.core (expand)"
2021/02/07 16:26:34 [TRACE] dag/walk: vertex "provider[\"registry.terraform.io/hashicorp/azurerm\"] (close)" is waiting for "azurerm_private_endpoint.keyvault-private-endpoint (expand)"
2021/02/07 16:26:34 [TRACE] dag/walk: vertex "meta.count-boundary (EachMode fixup)" is waiting for "azurerm_private_endpoint.keyvault-private-endpoint (expand)"
2021/02/07 16:26:34 [TRACE] dag/walk: vertex "root" is waiting for "meta.count-boundary (EachMode fixup)"
2021/02/07 16:26:38 [TRACE] dag/walk: vertex "azurerm_private_endpoint.keyvault-private-endpoint (expand)" is waiting for "azurerm_key_vault.core (expand)"
2021/02/07 16:26:39 [TRACE] dag/walk: vertex "meta.count-boundary (EachMode fixup)" is waiting for "azurerm_private_endpoint.keyvault-private-endpoint (expand)"
2021/02/07 16:26:39 [TRACE] dag/walk: vertex "root" is waiting for "meta.count-boundary (EachMode fixup)"
2021/02/07 16:26:39 [TRACE] dag/walk: vertex "provider[\"registry.terraform.io/hashicorp/azurerm\"] (close)" is waiting for "azurerm_private_endpoint.keyvault-private-endpoint (expand)"
2021/02/07 16:26:43 [TRACE] dag/walk: vertex "azurerm_private_endpoint.keyvault-private-endpoint (expand)" is waiting for "azurerm_key_vault.core (expand)"
2021/02/07 16:26:44 [TRACE] dag/walk: vertex "provider[\"registry.terraform.io/hashicorp/azurerm\"] (close)" is waiting for "azurerm_private_endpoint.keyvault-private-endpoint (expand)"
2021/02/07 16:26:44 [TRACE] dag/walk: vertex "meta.count-boundary (EachMode fixup)" is waiting for "azurerm_private_endpoint.keyvault-private-endpoint (expand)"
2021/02/07 16:26:44 [TRACE] dag/walk: vertex "root" is waiting for "meta.count-boundary (EachMode fixup)"
2021/02/07 16:26:47 [TRACE] vertex "azurerm_key_vault.core": visit complete
2021/02/07 16:26:47 [TRACE] vertex "azurerm_key_vault.core": dynamic subgraph encountered errors
2021/02/07 16:26:47 [TRACE] vertex "azurerm_key_vault.core": visit complete
2021/02/07 16:26:47 [TRACE] vertex "azurerm_key_vault.core (expand)": dynamic subgraph encountered errors
2021/02/07 16:26:47 [TRACE] vertex "azurerm_key_vault.core (expand)": visit complete
2021/02/07 16:26:47 [TRACE] dag/walk: upstream of "azurerm_private_endpoint.keyvault-private-endpoint (expand)" errored, so skipping
2021/02/07 16:26:47 [TRACE] dag/walk: upstream of "provider[\"registry.terraform.io/hashicorp/azurerm\"] (close)" errored, so skipping
2021/02/07 16:26:47 [TRACE] dag/walk: upstream of "meta.count-boundary (EachMode fixup)" errored, so skipping
2021/02/07 16:26:47 [TRACE] dag/walk: upstream of "root" errored, so skipping
2021/02/07 16:26:47 [TRACE] statemgr.Filesystem: removing lock metadata file .terraform.tfstate.lock.info

2021/02/07 16:26:47 [TRACE] statemgr.Filesystem: unlocking terraform.tfstate using fcntl flock
Error: retrieving `contact` for KeyVault: keyvault.BaseClient#GetCertificateContacts: Failure sending request: StatusCode=0 -- Original Error: context deadline exceeded

```
### Expected Behavior
Applied without error.

### Actual Behavior
On the second and future `terraform apply` operations get the following error:

```
Error: retrieving `contact` for KeyVault: keyvault.BaseClient#GetCertificateContacts: Failure sending request: StatusCode=0 -- Original Error: context deadline exceeded
```
### Steps to Reproduce

1. `terraform init`
2. `terraform apply`
3. `terraform apply`

**Comment by inforedaster** [comment]
i have the same problem but when i change Provisioning state to rejected into azure portal of a vault private endpoint and i retry terraform command it's works well

**Comment by cramsey86** [comment]
I am getting the same thing with Terraform version 0.13.5. I do not recall having this issue until integrating Private EndPoint as well

**Comment by cramsey86** [comment]
> I am getting the same thing with Terraform version 0.13.5. I do not recall having this issue until integrating Private EndPoint as well

So my issue was I had the Private EndPoint for the KeyVault configured in Terraform to the privatelink.vault.azure.net instead of privatelink.vaultcore.azure.net, so my agent was not resolving the keyvault correctly and getting that error.

**Comment by marrobi** [comment]
I "think" this could be caused by carrying out data plane operations as part of `resourceArmKeyVaultRead` - added in #8937 (thanks for the pointer @Tbohunek ).

https://github.com/terraform-providers/terraform-provider-azurerm/blob/65e37e28a2cf55f56e7dbc5f9c97044b3aa3ea14/azurerm/internal/services/keyvault/key_vault_resource.go#L671

This should only execute if contact is set. Maybe wrap in:

`if v, ok := d.GetOk("contact"); ok {`

`}`

@lawrencegripper appreciate a pointer, do you think I'm on the right track?

**Comment by tombuildsstuff** [comment]
@marrobi 

> This should only execute if contact is set

This needs to execute all the time to ensure the API and state match up (and will be an issue for the Read too - support for (conditionally) using the Resource Manager API's (as a feature-toggle) is blocked on the Service Team unfortunately.

**Comment by tombuildsstuff** [comment]
Support for that is tracked in #9738 - closing this in favour of that issue.

**Comment by marrobi** [comment]
Thanks, so what's the solution? Currently blocked with this issue.

Assistance appreciated.

**Comment by marrobi** [comment]
@tombuildsstuff not sure my issue is actually the same - I'm not blocking access to the public endpoint.

There is no deny on the KeyVault firewall.

Or does terraform behave differently if a private endpoint exists?

**Comment by marrobi** [comment]
@tombuildsstuff definitely a different issue to #9738 .

If I run 

```
az login --allow-no-subscriptions --service-principal --username $DEPLOY_CLIENT_ID --password $DEPLOY_CLIENT_SECRET --tenant $DEPLOY_TENANT_ID
az rest --method GET --uri "https://xxxx.vault.azure.net/certificates/contacts?api-version=7.1" --resource "https://vault.azure.net"
```
immediately before my TF apply, **I get a valid response**. The private endpoint isn't blocking access to the KeyVault.

If I deploy the KeyVault with ARM (hence removing the need to read the contact), I get the same issue with certificate read.

**Comment by magodo** [comment]
@marrobi Though this issue is marked as closed. While I just tried your above configuration (while with the latest provider) in my local PC (i.e. not accessible to the PE), it works fine for me (apply and next time plan/apply).

I noticed that you've created a private dns zone, while you didn't create the `azurerm_private_dns_zone_virtual_network_link`, which makes the DNS resolves to the key vault's public endpoint even resolving within the vnet that the PE resides in.

One thing happens on my side that will run into the exact same error as you saw is that if I enabled my corporation VPN on my local PC. In this case, the DNS resolving to the key vault endpoint will only return:

```shell
💤  dig +noall +answer mgdkv-core-uniquename.vault.azure.net
mgdkv-core-uniquename.vault.azure.net. 0 IN CNAME mgdkv-core-uniquename.privatelink.vaultcore.azure.net.
```

Terraform (actually the underlying SDK) will hang until context exceeded deadline. I'm wondering whether that is also your case?
